BOT-319 Sem search dead letter queue (DLQ)

wotbrew opened the pull request:
## Problem Statement

Semantic search uses a singleton indexer job with a durable watermark (provided in #61602), when a node dies, or an error occurs - the job always picks up the last watermark and continues from there.

This means if indexing fails due to some transient problem, like a connection error with the embedder - indexing will continue from where it left off once the issue resolves itself. 

But wait, what if, now here me out - the error **DOESN'T** resolve itself?

When any document in a batch persistently fails to index (due to malformed content, bug in our code for some search model case we haven't seen in test, embedding service issues with particular docs, etc.), the entire batch fails and the watermark cannot advance.

This creates a problem where a single poison document can indefinitely stall the indexer, preventing all subsequent documents from being processed.

Without a way of dealing with, you might need to patch metabase when the the problem is with _just one document or batch_, during which time the entire semantic search index becomes stale and you drown in `ERROR` logs.

## Solution

  This PR implements a dead letter queue (DLQ) system that ensures:

  - Problematic documents are moved to a separate DLQ table, allowing the main indexer watermark to progress
  - Failed documents are retried with shrinking batch sizes to isolate specific problem records, or issues that occur due to batch sizes being too big.
  - An exponential back-off policy is applied to retries - policies initially provided for 'more likely transient' and 'more likely permanent' categorizations of common errors (transient is preferred).
 
**TLDR;** The indexer can now continue processing new documents even when some fail repeatedly

## Stuff I changed

  `dlq.clj` 🆕 
  - New lib, companion for the `indexer` lib - DLQ writes and processing is orchestrated by the indexer (no data races!).
  - Per-index DLQ tables for tracking retry attempts and backoff timing for _that particular index_.
  - Functions for writing DLQ entries and processing them.

  `indexer.clj`
  - Added stall detection logic, activate DLQ when not getting anywhere
  - Cooperative DLQ draining during normal indexer operation (no additional job or thread).

  `index_metadata` schema changes
  - Added `indexer_stalled_at` timestamp to track stalls and apply the DLQ policy
  - Supports DLQ table lifecycle management (might need some fixes after rebase on latest mig commits).

Best bet for a detailed description on current policy and behaviour is to start the comments [here](https://github.com/metabase/metabase/pull/62070/files#diff-38673bb2c336c41192f3480090d1af7aef087a37d9e1ec98b5406f8285821937R138) in `indexer.clj` - the variables are likely to change as we see how things work in the real world.

wotbrew committed:
Add dead letter queue for the semantic search indexer

The DLQ system handles failed semantic search indexing operations if they are causing the indexer to stall.

- Applies back-off policies for retries (hard coded for now).
- Prevents batch poisoning from blocking indexer progress.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/dlq.clj:
```diff
@@ -0,0 +1,491 @@
+  "Attempts to index a batch of gate documents, separating upserts from deletes
+  processing them independently and then returning information on what failed and what succeeded.
+
+  Each operation swallows any error thrown by the indexer and performs, other than interrupts.
```
typo

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -44,45 +97,107 @@
+        ;; documents that prevent normal processing.
+        ;;
+        ;; DLQ Processing Strategy:
+        ;; - Any indexing failures are added to the dead letter queue with to be retried after the dlq/initial-backoff.
```
```suggestion
        ;; - Any indexing failures are added to the dead letter queue to be retried after the dlq/initial-backoff.
```

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -44,45 +97,107 @@
+          ;; once no longer failing, clear any stall from index_metadata
+          ;; this will cause the 'healthy' branch to be used again.
+          (when (empty? failures)
+            (log/info "Indexer recovered")
```
```suggestion
            (log/info "Indexer recovered from stalled state")
```

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -131,6 +253,83 @@
+  DLQ contains many entries ready for retry.
+
+  We do need to balance this for throughput during periods of prolonged (but transient) failure. it is
+  important that DLQ queue processing peak throughput is high when demand is high.
```
```suggestion
  important that DLQ processing peak throughput is high when demand is high.
```

appleby reviewed (approved):
Nice work! It's fun to wreak havoc and watch the DLQ kick in and recover from failures.

LGTM, but please [don't merge](https://metaboat.slack.com/archives/C07SJT1P0ET/p1755201287848099) until after the feature branch lands in master.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/dlq.clj:
```diff
@@ -0,0 +1,491 @@
+  ^Duration [^long retry-count policy]
+  (let [{:keys [^Duration cap]} policy
+        cap-ms (.toMillis cap)
+        delay  (min (max-delay-ms retry-count policy) cap-ms)]
```
Yep I deleted the jitter by mistake! Tired vim arms :) - Good catch.

wotbrew committed:
Add the jitter factor I mistakenly deleted

wotbrew committed:
Doc fixes

wotbrew committed:
Slightly better stall recovery log

wotbrew committed:
Kondo

wotbrew committed:
Index attempt_at on DLQ table

wotbrew committed:
Add some more logging

wotbrew commented:
Note: Post rebase today I've noticed need to make a couple tweaks to init (and add a test), due to the new migrations PR landing.

wotbrew committed:
Init includes DLQ creation

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/pgvector_api.clj:
```diff
@@ -31,25 +32,29 @@
-          (semantic.index-metadata/activate-index! tx index-metadata index-id)
-          index)
-        active-index))))
+  (let [{active-index :index
```
@bronsa is also in this area and need to make sure logic here makes sense for reinit also

wotbrew committed:
Fix some rebase collisions

wotbrew committed:
Formatting

wotbrew committed:
Add some asserts to show dlq creation during init

wotbrew committed:
Name test fixed with hashing

wotbrew committed:
Inline table-name-prefix

wotbrew committed:
Bump index meta version

wotbrew committed:
Prefer hex to b64 for table/index names they remain usable unescaped.

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/index.clj:
```diff
@@ -213,33 +215,37 @@
-  [embedding-model]
-  (let [table-name (model-table-name embedding-model)]
+  [embedding-model & {:keys [table-name]}]
+  (let [table-name (or table-name (model-table-name embedding-model))]
```
The random suffix no longer happens for each call (made tests etc quite complex to think about) - it is predictable as before, in `pgvector-api/initialize-index!` we now use this option to vary the name al-carte when doing a `:force-reset? true`.

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/index.clj:
```diff
@@ -213,33 +215,37 @@
-                    {:table-name resource-name
-                     :length (count resource-name)
-                     :limit 63}))))
+(defn hash-identifier-if-exceeds-pg-limit
```
Hashing seems ok to me, and safer than throwing. My lizard brain would still feel better if I got an un-ignorable slap in the face in case I accidentally add a new table that exceeds the limit (don't really care about index names), but admittedly it's annoying for tests that just want to add some `"mock_blah_"` prefix that winds up pushing it over the limit, and I think I'd rather it hash than throw if someone switches to a random new embedding model with a long name.

wotbrew replied:
Yes that would be cool but I was already pushing it with adding it to this PR and didn't want to write too many fancy tests! Lets follow up with that as an improvement? I think the names are very valuable for debugging and we should still strive to be in a situation where hashing is used as a last resort (i.e ideally never) in prod.

noahmoss reviewed (approved):
👍 The doc strings & comments are all super helpful so thanks for that. Did a more focused review of the core DLQ implementation and only had a few minor comments

wotbrew commented on enterprise/backend/src/metabase_enterprise/semantic_search/dlq.clj:
```diff
@@ -0,0 +1,518 @@
+                                        :last_attempted_at :excluded.last_attempted_at}
+                               ;; condition: if exiting dlq entry reflects a more recent gate failure
+                               ;; prefer the new attempt_at / retry_count from the gate to the prior DLQ entry.
+                               :where  [:<= (keyword (name (dlq-table-name-kw index-metadata index-id)) "error_gated_at")
```
redundant hang over for sure 👍

appleby committed:
Fix next-dlq-run test in on-indexing-idle

The test was backwards, so the loop only idled when `next-dlq-run` was due, rather than avoiding idle in that case.

appleby committed:
Ensure exit-early-due-to-cold-data? even if we have no last-seen-change

Ensure we still yield back to quartz if more than exit-early-cold-duration seconds have elapsed since the loop start
and we have no last-seen-change. This can happen in cases where no docs were updated since the loop start if our
watermark is already at the end of the gate table.

appleby committed:
Reduce log level to :trace for "nothing to do" case in indexing-step

appleby committed:
Whitespace cleanup

appleby committed:
Reduce ee-search-indexer-exit-early-cold-duration from 30s to 5s

30s is longer than the quartz interval for the indexer job, so we'll just get rescheduled immediately after yielding
control.

appleby committed:
Increase the run-frequency for the SemanticSearchIndexer job from every 10s to every 20s

appleby committed:
Move table-exists? into shared util ns and reuse

appleby committed:
Simplify honeysql qualified col reference in add-entries!

appleby committed:
Fix comment typo: s/indexes/indexer/

appleby committed:
Add missing seq in try-batch!

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -92,9 +222,16 @@
+
     (cond
+      ;; if the next DLQ run should happen on the next iteration, we should not idle
+      (and next-dlq-run (.isAfter (.instant clock) next-dlq-run))
```
order was swapped here: ensure we *do* idle unless `next-dlq-run` is ready

https://github.com/metabase/metabase/pull/62070/commits/69839a5d210567e52ac96657440c84506abda8be

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/indexer.clj:
```diff
@@ -128,9 +265,94 @@
+           (.isAfter now (.plus last-seen-change exit-early-cold-duration)))
+      (and (nil? last-seen-change)
+           start-time
+           (.isAfter now (.plus start-time exit-early-cold-duration)))))
```
changed this so that we also yield back to quartz if we haven't seen any changes since the start.

https://github.com/metabase/metabase/pull/62070/commits/a77aeb2a0fec9c7134e045717be0d206ff53cb43

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/settings.clj:
```diff
@@ -114,7 +114,7 @@
   "Number of seconds indexer should wait to see new data before yielding back to quartz."
   :type :integer
-  :default 30
+  :default 5
```
30s is longer than the quartz interval for the indexer job, so we were just get rescheduled immediately after yielding
control from `indexing-loop`.

appleby replied:
oops, had to revert this. this value needs to be greater than the `dlq-frequency`. this does mean we get rescheduled immediately when we yield back to quartz, so possibly could tweak these later, but for now just revert.

appleby commented on enterprise/backend/src/metabase_enterprise/semantic_search/task/indexer.clj:
```diff
@@ -53,7 +53,7 @@
 
 (def ^:private ^Duration startup-delay (Duration/parse "PT10S"))
-(def ^:private ^Duration run-frequency (Duration/parse "PT10S"))
+(def ^:private ^Duration run-frequency (Duration/parse "PT20S"))
```
Potentially controversial, but this felt like a better pacing to me. Our idle sleep time is 3s and the cold exit timeout is now 5s, so in the idle state we sleep 3s, sleep 3s, yield to quartz, sleep another 10-6 = 4s, then get rescheduled. So the drum beat was pretty steady, and since there was little difference between the idle sleep and the reschedule sleep, you might as well not yield control at that point (which could also be an option). With the new values, we sleep 3s, 3s, then 14s before getting rescheduled, which still feels fairly responsive.

appleby replied:
^ reverted the change to the cold exit timer, so now we sleep 10 x 3s, yield to quartz, then get immediately rescheduled.

wotbrew replied:
Yes I think this is a good change and a better default for now - I expected to balance all the related variables using the metrics (coming soon) on idle/busy times, indexing delays etc. The values that are in there are literally first guesses!

appleby commented:
@wotbrew pushed a fix per slack discussion earlier, and went ahead and resolved @noahmoss' comments as well. let me know what you think when you're back online and feel free to merge if everything looks ok

appleby committed:
Change ee-search-indexer-exit-early-cold-duration back to 30s

This value needs to be higher than the dlq-frequency, otherwise the dlq never runs.

wotbrew commented:
All good @appleby thanks so much, looks good to merge!
